### PR TITLE
HAL APIs backwards compatibility support.

### PR DESCRIPTION
**Pull request by XuGuohui**
### Problem

Older user application that invokes HAL APIs directly compile against new Device OS will fail.
 
### Solution

Add inlined old HAL APIs with deprecated attribute.

### Steps to Test

Build `user/tests/wiring/api`.

---

### Completeness

- [x] User is totes amazing for contributing!
- [x] Contributor has signed CLA ([Info here](https://github.com/spark/firmware/blob/develop/CONTRIBUTING.md))
- [x] Problem and Solution clearly stated
- [ ] Run unit/integration/application tests on device
- [ ] Added documentation
- [ ] Added to CHANGELOG.md after merging (add links to docs and issues)

### CODE REVIEW

File: build/module-defaults.mk [hunk 1]
```diff
@@ -37,7 +37,7 @@ ifeq ($(USE_SWD_JTAG),y)
 COMPILE_LTO ?= n
 endif
 
-WARNINGS_AS_ERRORS ?= y
+WARNINGS_AS_ERRORS ?= n
```
Comment (avtolstoy) [inline build/module-defaults.mk:40]: This should be left as-is. Only the application is allowed to be built with just warnings.

File: hal/inc/adc_hal.h [hunk 2]
```diff
@@ -42,6 +42,11 @@ int32_t hal_adc_read(pin_t pin);
 void hal_adc_dma_init();
 int hal_adc_sleep(bool sleep, void* reserved);
 
+
+// Deprecated *dynalib* APIs for backwards compatibility
+void inline __attribute__((always_inline, deprecated("Use hal_adc_set_sample_time() instead"))) HAL_ADC_Set_Sample_Time(uint8_t sample_time) { hal_adc_set_sample_time(sample_time); }
```
Comment (avtolstoy) [inline hal/inc/adc_hal.h:47]: Let's move to a separate `adc_hal_compat.h` header. Same for all the other headers.

File: hal/inc/usart_hal.h [hunk 3]
```diff
@@ -176,6 +176,24 @@ ssize_t hal_usart_write_buffer(hal_usart_interface_t serial, const void* buffer,
...
+void inline __attribute__((always_inline, deprecated("Use hal_usart_half_duplex() instead"))) HAL_USART_Half_Duplex(HAL_USART_Serial serial, bool Enable) { hal_usart_half_duplex(serial, Enable); }
+void inline __attribute__((always_inline, deprecated("Use hal_usart_begin_config() instead"))) HAL_USART_BeginConfig(HAL_USART_Serial serial, uint32_t baud, uint32_t config, void* reserved) { hal_usart_begin_config(serial, baud, config, reserved); }
+uint32_t inline __attribute__((always_inline, deprecated("Use hal_usart_write_nine_bits() instead"))) HAL_USART_Write_NineBitData(HAL_USART_Serial serial, uint16_t data) { return hal_usart_write_nine_bits(serial, data); }
+void inline __attribute__((always_inline, deprecated("Use hal_usart_send_break() instead"))) HAL_USART_Send_Break(HAL_USART_Serial serial, void* reserved) { hal_usart_send_break(serial, reserved); }
+uint8_t inline __attribute__((always_inline, deprecated("Use hal_usart_break_detected() instead"))) HAL_USART_Break_Detected(HAL_USART_Serial serial) { return hal_usart_break_detected(serial); }
+
```
Comment (avtolstoy) [inline hal/inc/usart_hal.h:196]: API tests `wiring/api` should be added for all these compat functions.

File: hal/inc/usart_hal.h [hunk 4]
```diff
@@ -176,6 +176,24 @@ ssize_t hal_usart_write_buffer(hal_usart_interface_t serial, const void* buffer,
 ssize_t hal_usart_read_buffer(hal_usart_interface_t serial, void* buffer, size_t size, size_t elementSize);
 ssize_t hal_usart_peak_buffer(hal_usart_interface_t serial, void* buffer, size_t size, size_t elementSize);
 
+
+// Deprecated *dynalib* APIs for backwards compatibility
+void inline __attribute__((always_inline, deprecated("Use hal_usart_init() instead"))) HAL_USART_Init(HAL_USART_Serial serial, Ring_Buffer *rx_buffer, Ring_Buffer *tx_buffer) { hal_usart_init(serial, rx_buffer, tx_buffer); }
```
Comment (avtolstoy) [inline hal/inc/usart_hal.h:181]: Nit: let's format all these functions normally on 3 lines.

Comment (avtolstoy) [inline hal/inc/usart_hal.h:181]: Nit: `always_inline` is probably redundant. The compiler will easily figure it out on its own.

Comment (avtolstoy) [reply hal/inc/usart_hal.h:181]: `inline return_type`, not `return_type inline`.

File: user/tests/wiring/api/spi.cpp [hunk 5]
```diff
@@ -199,3 +199,26 @@ test(spi_lock)
...
+    API_COMPILE(HAL_SPI_End(HAL_SPI_INTERFACE1));
+    API_COMPILE(HAL_SPI_Set_Bit_Order(HAL_SPI_INTERFACE1, 0));
+    API_COMPILE(HAL_SPI_Set_Data_Mode(HAL_SPI_INTERFACE1, 0));
+    API_COMPILE(HAL_SPI_Set_Clock_Divider(HAL_SPI_INTERFACE1, 0));
+    API_COMPILE(HAL_SPI_Send_Receive_Data(HAL_SPI_INTERFACE1, 0));
+    API_COMPILE(HAL_SPI_DMA_Transfer(HAL_SPI_INTERFACE1, NULL, NULL, 0, NULL));
+    API_COMPILE(HAL_SPI_Is_Enabled_Old());
+    API_COMPILE(HAL_SPI_Is_Enabled(HAL_SPI_INTERFACE1));
+    API_COMPILE(HAL_SPI_Info(HAL_SPI_INTERFACE1, NULL, NULL));
+    API_COMPILE(HAL_SPI_Set_Callback_On_Select(HAL_SPI_INTERFACE1, NULL, NULL));
```
Comment (avtolstoy) [inline user/tests/wiring/api/spi.cpp:218]: Compat types are missing.

File: hal/inc/spi_hal.h [hunk 6]
```diff
@@ -55,6 +55,8 @@ typedef enum hal_spi_state_t {
 
 typedef void (*hal_spi_dma_user_callback)(void);
 typedef void (*hal_spi_select_user_callback)(uint8_t);
+typedef hal_spi_dma_user_callback HAL_SPI_DMA_UserCallback; // For backwards compatibility only
```
Comment (avtolstoy) [inline hal/inc/spi_hal.h:58]: These can now also be moved into compat header. We can also put deprecated attribute on them.

File: hal/inc/adc_hal_compat.h [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHAN'TABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef __ADC_HAL_COMPAT_H
```
Comment (avtolstoy) [inline hal/inc/adc_hal_compat.h:18]: Nit: anything beginning with `__` is reserved for stdlib/compiler usage.

**Comment by XuGuohui** [comment]
The Travis CI build failed due to treating warnings as errors.

**Pull request merged**